### PR TITLE
ENH added PyDMAnalogIndicator based on the PyDMScaleIndicator

### PR DESCRIPTION
**Pull request by Ryan-McClanahan**
PyDMAnalogIndicator is a redesign of the PyDMScaleIndIndicator.  It adds alarm regions into the scale.  The alarm regions indicate minor and major alarm values and change color when the indicator points at them.  Changed the indicator shape to a triangle and moved it outside the scale.  The major and minor alarm values can be assigned by the user or pulled from the meta data of the PV.

Future work
Adding set point indicator. Requires use of second channel
Add polish

### CODE REVIEW

File: pydm/widgets/analog_indicator.py [hunk 1]
```diff
@@ -0,0 +1,1576 @@
+from PyQt5.QtCore import QPointF
+
+from .base import PyDMWidget, TextFormatter
+from qtpy.QtGui import QColor, QPolygon, QPen, QPainter
+from qtpy.QtWidgets import QFrame, QVBoxLayout, QHBoxLayout, QLabel, QSizePolicy, QWidget, QGridLayout
+from qtpy.QtCore import Qt, QPoint, Property
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:6]: It's usually best practice to import from `qtpy` instead of directly from `PyQt5`. This will redirect the import to `PyQt5` if that's what the user has installed, but will also help with cross-support to making your widget work in `PySide2` and other qt python bindings.

```suggestion
from .base import PyDMWidget, TextFormatter
from qtpy.QtGui import QColor, QPolygon, QPen, QPainter
from qtpy.QtWidgets import QFrame, QVBoxLayout, QHBoxLayout, QLabel, QSizePolicy, QWidget, QGridLayout
from qtpy.QtCore import Qt, QPoint, QPointF, Property
```

File: pydm/widgets/analog_indicator.py [hunk 2]
```diff
@@ -0,0 +1,1576 @@
...
+from qtpy.QtWidgets import QWIDGETSIZE_MAX
+from .channel import PyDMChannel
+import sys
+
+
+class QScale(QFrame):
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:12]: Are you using the `QFrame` inheritance to let the user adjust the frame style? I don't see anything in the `QScale` class specifically that references it, so it isn't obvious to me why this isn't just a `QWidget`.

File: pydm/widgets/analog_indicator.py [hunk 3]
```diff
@@ -0,0 +1,1576 @@
...
+        Whether or not the scale indicator should use the minor alarm information
+        from the channel.
+
+        Parameters
+        ----------
+        checked : bool
+            True to use the major alarms from the Channel, False to use the user-defined
+            values.
+        """
+        print("_major_alarm_from_channel")
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:1492]: You've accidentally left `print` statements all over this setter

File: pydm/widgets/analog_indicator.py [hunk 4]
```diff
@@ -0,0 +1,1576 @@
...
+        value : float
+            The new lower major alarm value.
+        """
+
+        if self._major_alarm_from_channel:
+            return
+
+        self._user_upper_major_alarm = value
+        self.scale_indicator.set_upper_major_alarm(self._user_upper_major_alarm)
+        #self.update_labels()
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:1547]:
```suggestion
```

File: pydm/widgets/analog_indicator.py [hunk 5]
```diff
@@ -0,0 +1,1576 @@
...
+        value : float
+            The new lower major alarm value.
+        """
+
+        if self._major_alarm_from_channel:
+            return
+
+        self._user_lower_major_alarm = value
+        self.scale_indicator.set_lower_major_alarm(self._user_lower_major_alarm)
+        #self.update_labels()
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:1576]:
```suggestion
```

File: pydm/widgets/qtplugins.py [hunk 6]
```diff
@@ -241,6 +242,13 @@
                                             icon=ifont.icon("level-up-alt")
                                             )
 
+# Analog Indicator plugin
+PyDMAnalogIndicatorPlugin = qtplugin_factory(PyDMAnalogIndicator,
+                                            group=WidgetCategory.DISPLAY,
+                                            extensions=BASE_EXTENSIONS,
+                                            icon=ifont.icon("level-up-alt")
+                                            )
```
Comment (ZLLentz) [inline pydm/widgets/qtplugins.py:250]: Slight indent mismatch
```suggestion
PyDMAnalogIndicatorPlugin = qtplugin_factory(PyDMAnalogIndicator,
                                             group=WidgetCategory.DISPLAY,
                                             extensions=BASE_EXTENSIONS,
                                             icon=ifont.icon("level-up-alt")
                                             )
```

File: pydm/widgets/analog_indicator.py [hunk 7]
```diff
@@ -0,0 +1,1576 @@
...
+        ----------
+        new_major_alarm : float
+            The lower major alarm of the scale.
+        """
+        self._lower_major_alarm = new_major_alarm
+
+    def get_show_ticks(self):
+        return self._show_ticks
+
+    def set_show_ticks(self, checked):
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:395]: Seems like you stopped writing docstrings for this class at about here in the file- I don't blame you though, you've got a lot of method here!

**Review by ZLLentz — COMMENTED** [review]
I went through this line by line to try to catch any oddities- overall, code health looks good! You've added a lot of methods here, it's hard to find any that weren't needed. Some of the code paths are hard to follow with the inheritance tree but that's fine.

I'm going to run this locally now and check it out in action before returning with part 2 of this review.

I highly recommend trying to run `flake8` locally on this file to see what comes up- lots of small style things can be fixed pretty easily!

File: pydm/widgets/analog_indicator.py [hunk 8]
```diff
@@ -0,0 +1,1576 @@
...
+        """
+        Calculate the position (pixel) in which the pointer should be drawn for a given value.
+        """
+        if value is None or value < self._lower_limit or value > self._upper_limit or \
+           self._upper_limit - self._lower_limit == 0:
+            proportion = -1 # Invalid
+        else:
+            proportion = (value - self._lower_limit) / (self._upper_limit - self._lower_limit)
+
+        position = int(proportion * self._widget_width)
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:294]: I ran into this problem while taking it for a spin. I dragged the widget into an empty designer ui. I added a channel to an analog PV. The widget had no drawn bar so I went to make it bigger, and then this happened before it could render the bar:
```
Exception occurred while running Qt Designer.
Traceback (most recent call last):
  File "/cds/home/z/zlentz/pydev/pydm/widgets/analog_indicator.py", line 277, in paintEvent
    self.draw_minor_alarm_region()
  File "/cds/home/z/zlentz/pydev/pydm/widgets/analog_indicator.py", line 184, in draw_minor_alarm_region
    lower_minor_alarm_width = self.calculate_position_for_value(self._lower_minor_alarm)
  File "/cds/home/z/zlentz/pydev/pydm/widgets/analog_indicator.py", line 294, in calculate_position_for_value
    position = int(proportion * self._widget_width)
ValueError: cannot convert float NaN to integer
```

Comment (ZLLentz) [reply pydm/widgets/analog_indicator.py:294]: This crashes the entire designer session very dramatically btw

Comment (ZLLentz) [reply pydm/widgets/analog_indicator.py:294]: I switched to a different PV and it was fine- the PV that failed had the following hilarious metadata. I suggest finding a way to handle these in a non-crashing way:
```
$ caget -d 30 XPP:USR:MMS:01.RBV
XPP:USR:MMS:01.RBV
    Native data type: DBF_DOUBLE
    Request type:     DBR_CTRL_FLOAT
    Element count:    1
    Value:            -70.3204
    Status:           COMM
    Severity:         INVALID
    Units:            mm
    Precision:        4
    Lo disp limit:    -50
    Hi disp limit:    150
    Lo alarm limit:   nan
    Lo warn limit:    nan
    Hi warn limit:    nan
    Hi alarm limit:   nan
    Lo ctrl limit:    -50
    Hi ctrl limit:    150
```

File: pydm/widgets/analog_indicator.py [hunk 9]
```diff
@@ -0,0 +1,1576 @@
...
+                    self.scale_indicator.set_upper_limit(self._upper_ctrl_limit)
+            else:
+                self.scale_indicator.set_lower_limit(self._user_lower_limit)
+                self.scale_indicator.set_upper_limit(self._user_upper_limit)
+            self.update_labels()
+
+
+
+    @Property(float)
+    def userUpperLimit(self):
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:1319]: Minor designer usability thing: since all of the "upper" limits are defined before the "lower" limits, they appear in the order of upper first, lower second in the designer editor. So if I want a range from 30 to 50, for example, my designer editor looks like:
```
userUpperLimit         50
userLowerLimit         30
```
which reads out in my brain as "from 50 to 30" and breaks me a little. I think the other order would be more intuitive so I could read down "from low to high" in the settings.

File: pydm/widgets/analog_indicator.py [hunk 10]
```diff
@@ -0,0 +1,1576 @@
...
+        self.draw_ticks() -> draws scale ticks
+        self.draw_set_point -> draws diamond in the middle of the scale. can use current self.draw_indicator() for this
+        """
+        self.draw_background()
+        self.draw_ticks()
+
+        self.draw_minor_alarm_region()
+        self.draw_major_alarm_region()
+        self.draw_indicator()
+        #self.draw_set_point()
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:280]:
```suggestion
```

File: pydm/widgets/analog_indicator.py [hunk 11]
```diff
@@ -0,0 +1,1576 @@
...
+
+        self._painter.end()
+
+    def calculate_position_for_value(self, value):
+        """
+        Calculate the position (pixel) in which the pointer should be drawn for a given value.
+        """
+        if value is None or value < self._lower_limit or value > self._upper_limit or \
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:288]: This may give strange results when a user mistakenly puts in lower limit as the high value and upper limit as the low value. Perhaps that should be handled here, or maybe there should be some input validation on the upper/lower limits setters.

**Comment by klauer** [comment]
Screenshots for the slides:
image
image
image

File: pydm/widgets/analog_indicator.py [hunk 12]
```diff
@@ -0,0 +1,1576 @@
...
+
+    def set_inverted_appearance(self, inverted):
+        self._inverted_appearance = inverted
+        self.adjust_transformation()
+        self.repaint()
+
+    def get_bar_indicator(self):
+        return self._barIndicator
+
+    def set_bar_indicator(self, checked):
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:427]: I couldn't figure out what the bar indicator setting did- everything else worked as expected!

File: pydm/widgets/analog_indicator.py (hunk 8 above)
Comment (Ryan-McClanahan) [reply pydm/widgets/analog_indicator.py:302]: Hi Zach, thanks for all of your detailed feedback! 
I want to confirm that this crash is in designer and not a live PyDM display?  I want to try to replicate and fix this, however, I'm not sure what would cause a crash inside of designer as the alarm meta data should not be being pulled in in a designer session.
Thanks again!

Comment (ZLLentz) [reply pydm/widgets/analog_indicator.py:302]: Yes, this crash occurs in designer. I haven't checked normal pydm yet, I was fixated on the editor crash.

> the alarm meta data should not be being pulled in in a designer session

Are you sure about this? We get the live values in designer to help the user see what their widget looks like using real data, so I assumed that all the metadata was expected too.

File: pydm/widgets/analog_indicator.py [hunk 13]
```diff
@@ -0,0 +1,1584 @@
...
+        if self._value < self._lower_minor_alarm and self._value > self._lower_major_alarm:
+            self._painter.setBrush(self._minor_alarm_color)
+        else:
+            self._painter.setBrush(self._minor_alarm_region_color)
+
+        self._painter.drawRect(0, pointer_height, lower_minor_alarm_width, minor_alarm_height)
+        """
+        sets the pen color to alarm if the value is in the upper minor alarm region
+        """
+        if self._value > self._upper_minor_alarm and self._value < self._upper_major_alarm:
```
Comment (ZLLentz) [inline pydm/widgets/analog_indicator.py:208]: No idea how to get to this codepath, but it indicates that some care should be taken with using user input in draw functions
```
Exception occurred while running Qt Designer.
Traceback (most recent call last):
  File "/cds/home/z/zlentz/pydev/pydm/widgets/analog_indicator.py", line 284, in paintEvent
    self.draw_minor_alarm_region()
  File "/cds/home/z/zlentz/pydev/pydm/widgets/analog_indicator.py", line 208, in draw_minor_alarm_region
    if self._value > self._upper_minor_alarm and self._value < self._upper_major_alarm:
TypeError: '<' not supported between instances of 'int' and 'NoneType'
```

File: pydm/widgets/analog_indicator.py (hunk 8 above)
Comment (Ryan-McClanahan) [reply pydm/widgets/analog_indicator.py]: Fixed!

**Review by YektaY — APPROVED** [review]
Looks good!
